onnxruntime/core/providers/webnn/builders/impl/gqa_op_builder.cc

```diff
@@ -0,0 +1,542 @@
+    we cast them to float32 to ensure data precision.
+
+          query      key               value
+            |         |                  |
+         Reshape   Reshape            Reshape (B,S,H,N)     seqlens_k
+            |         |                  |                  /       |
+            |         |       past_value |   (scatter_indices*)     |
+        q_Transpose   |              \   |   /                      |
+        (0,2,1,3)     | past_key    ScatterND-----------------------|------> present_value
+             \        |  /              |                           |
+present_key<--\----ScatterND         Expand(G)      (attention_bias, one/finfo_min mask*)
+               \      |                 |              /
+               |   Expand(G)            |             /
+               |      |                 |            /
+               |  k_Transpose           |           /
+               |   (0,1,3,2)            |          /
+               |      |                 |         /
+            +---------------------------------------+
+            |        ScaledDotProductAttention      |
+            +---------------------------------------+
+                             |
+                           output
+*/
+
+Status GroupQueryAttentionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                           const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  emscripten::val query_input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val key_input = model_builder.GetOperand(input_defs[1]->Name());
+  emscripten::val value_input = model_builder.GetOperand(input_defs[2]->Name());
+  emscripten::val past_key_input = model_builder.GetOperand(input_defs[3]->Name());
+  emscripten::val past_value_input = model_builder.GetOperand(input_defs[4]->Name());
+  emscripten::val seqlens_k_input = model_builder.GetOperand(input_defs[5]->Name());
+
+  std::vector<int64_t> input_q_shape, input_past_k_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_q_shape, logger), "Cannot get query shape");
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[3], input_past_k_shape, logger), "Cannot get past_key shape");
+
+  NodeAttrHelper helper(node);
+  const uint32_t kv_num_heads = helper.Get("kv_num_heads", 0);
+  const uint32_t num_heads = helper.Get("num_heads", 0);
+
+  const uint32_t batch_size = SafeInt<uint32_t>(input_q_shape[0]);
+  const uint32_t qkv_sequence_length = SafeInt<uint32_t>(input_q_shape[1]);
+  const uint32_t qkv_hidden_size = SafeInt<uint32_t>(input_q_shape[2]);
+  const uint32_t head_size = SafeInt<uint32_t>(qkv_hidden_size / num_heads);
+  const uint32_t past_sequence_length = SafeInt<uint32_t>(input_past_k_shape[2]);
+  const uint32_t group_size = SafeInt<uint32_t>(num_heads / kv_num_heads);
+
+  const float scale_value = helper.Get("scale", 1 / static_cast<float>(sqrt(head_size)));
```

Comment:
This would call the integer overload of sqrt root. Is that desired? Do we want to convert to float first? e.g. `sqrt(float(head_size))`. (I don't know the answer) Can 0 reach here if `qkv_hidden_size / num_heads` if the caller passed bad shapes, or if the model contained dubious values? Maybe there are checks elsewhere, but we should ensure that a bad Status is returned rather than undefined behavior (or a crash if a bad value gets propagated).